org/mozilla/jss/ssl/javax/JSSEngineReferenceImpl.java

```diff
@@ -0,0 +1,1170 @@
+        }
+
+        // wire_data is the number of bytes written to dst. This is bounded
+        // above by two fields: the number of bytes we can read from
+        // write_buf, and the size of dst, if present.
+        int wire_data = (int) Buffer.ReadCapacity(write_buf);
+        if (dst == null) {
+            wire_data = 0;
+        } else {
+            // We want to know how much free space there is in dst for us to
+            // write to. This is given by dst.remaining().
+            wire_data = Math.min(wire_data, dst.remaining());
+        }
+
+        // Try reading data from write_buf to dst
+        if (wire_data > 0) {
+            byte[] wire_buffer = Buffer.Read(write_buf, wire_data);
+            dst.put(wire_buffer);
+            debug("JSSEngine.wrap() - Wrote " + wire_buffer.length + " bytes to dst.");
+        }
+
+        logWrap(dst);
+
+        // Before we return, check if an exception occurred and throw it if
+        // one did.
+        if (ssl_exception != null) {
+            info("JSSEngine.wrap() - Got SSLException: " + ssl_exception);
+            SSLException excpt = ssl_exception;
+            ssl_exception = null;
+            cleanup();
+            throw excpt;
+        }
+
+        // Need a way to introspect the open/closed state of the TLS
+        // connection.
+
+        SSLEngineResult.Status handshake_status = SSLEngineResult.Status.OK;
+
+        if (ssl_exception == null && seen_exception) {
+            debug("Seen and processed exception; closing inbound and outbound because this was the last wrap(...)");
+            closeInbound();
+            closeOutbound();
+        }
+
+        if (is_inbound_closed && is_outbound_closed) {
+            debug("Socket is currently closed.");
+            handshake_status = SSLEngineResult.Status.CLOSED;
+        }
+
+        debug("JSSEngine.wrap() - Finished");
+        debug(" - Status: " + handshake_status);
+        debug(" - Handshake State: " + handshake_state);
+        debug(" - wire_data: " + wire_data);
+        debug(" - app_data: " + app_data);
+
+        tryCleanup();
+        return new SSLEngineResult(handshake_status, handshake_state, app_data, wire_data);
+    }
+
+    /**
+     * Calls cleanup only if both inbound and outbound data streams are
+     * closed.
+     *
+     * This prevents accidental cleanup in the case of a partially open
+     * connection.
+     */
+    public void tryCleanup() {
+        debug("JSSEngine: tryCleanup()");
+        if (is_inbound_closed && is_outbound_closed) {
+            // throw new RuntimeException("Probably shouldn't be here!");
+            cleanup();
+        }
+    }
+
+    /**
+     * Performs cleanup of internal data, closing both inbound and outbound
+     * data streams if still open.
+     */
+    public void cleanup() {
+        debug("JSSEngine: cleanup()");
+
```

Comment:
Yes. We give two options:

 - `tryCleanup()` is always safe to call. If both sides of the connection are closed, it is safe to clean up after ourselves, because no further communication is expected over this SSLEngine socket. It in turn calls `cleanup()` when those conditions hold, else is a no-op.
 - `cleanup()` forces a cleanup, regardless of state. This should be used by the GC, and in cases where we know the connection terminated (timeout, ...). This actually frees the memory &c. 

There's no "spec" here, hence why I made both a try and a do-er. :-)

Reply:
Maybe I should formalize this into a specification before we get too far... Hmm...